lib.php

```diff
@@ -25,133 +25,129 @@
 defined('MOODLE_INTERNAL') || die;
 
 function tool_abconfig_after_config() {
-    global $CFG, $DB, $SESSION;
+    try {
+        global $CFG, $SESSION;
 
-    // Check if the param to disable ABconfig is present, if so, exit
-    if (array_key_exists('abconfig', $_GET) && $_GET['abconfig'] == 'off') {
-        if (is_siteadmin()) {
-            return null;
-        }
-    }
+        // Setup experiment manager
+        $manager = new tool_abconfig_experiment_manager();
 
-    // Check URL params, and fire any experiments in the params
-    foreach ($_GET as $experiment => $condition) {
-        $excompare = $DB->sql_compare_text($experiment, strlen($experiment));
-        $exrecord = $DB->get_record_sql("SELECT * FROM {tool_abconfig_experiment} WHERE shortname = ?", array($excompare));
+        // Check if the param to disable ABconfig is present, if so, exit
+        if (array_key_exists('abconfig', $_GET) && $_GET['abconfig'] == 'off') {
+            if (is_siteadmin()) {
+                return null;
+            }
+        }
 
-        if (!empty($exrecord)) {
-            $condcompare = $DB->sql_compare_text($condition, strlen($condition));
-            $condrecord = $DB->get_record_sql("SELECT * FROM {tool_abconfig_condition} WHERE condset = ? and experiment = ?", array($condcompare, $exrecord->id));
-            if (!empty($condrecord)) {
-                tool_abconfig_execute_command_array($condrecord->commands, $exrecord->shortname);
+        // Check URL params, and fire any experiments in the params
+        foreach ($_GET as $experiment => $condition) {
+            // Get all experiments
+            $experiments = $manager->get_experiments();
```

Comment:
I think this should be outside the loop so it's only called once. it is cached, but still nicer.

It's a separate issue but this could be a potential source of weird hard to track down bugs, like if you accidentally called a conditionset 'id' then it would trigger all the time. Give that this is only for testing, I think easiest is to only do all this if you are admin. Normal users should not be able to control their own experiment conditions